src/core/eme/handle_encrypted_event.ts

```diff
@@ -0,0 +1,127 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Observable } from "rxjs/Observable";
+import { IMediaKeySession } from "../../compat";
+import config from "../../config";
+import { EncryptedMediaError } from "../../errors";
+import log from "../../utils/log";
+import createSession from "./create_session";
+import { IMediaKeysInfos } from "./types";
+import InitDataStore from "./utils/init_data_store";
+import isSessionUsable from "./utils/is_session_usable";
+
+export interface IHandledEncryptedEvent {
+  type : "created-session" |
+    "loaded-open-session" |
+    "loaded-persistent-session";
+  value : {
+    mediaKeySession : MediaKeySession|IMediaKeySession;
+    sessionType : MediaKeySessionType;
+    initData : Uint8Array; // assiociated initialization data
+    initDataType : string; // type of the associated initialization data
+  };
+}
+
+const {
+  EME_MAX_SIMULTANEOUS_MEDIA_KEY_SESSIONS: MAX_SESSIONS,
+} = config;
+
+/**
+ * Handle MediaEncryptedEvents sent by a HTMLMediaElement:
+ * Either create a session, skip the event if it is already handled or
+ * recuperate a previous session and returns it.
+ * @param {Event} encryptedEvent
+ * @param {Object} handledInitData
+ * @param {Object} mediaKeysInfos
+ * @returns {Observable}
+ */
+export default function handleEncryptedEvent(
+  encryptedEvent : MediaEncryptedEvent,
+  handledInitData : InitDataStore,
+  mediaKeysInfos : IMediaKeysInfos
+) : Observable<IHandledEncryptedEvent> {
+  return Observable.defer(() => {
+    const {
+      initDataType,
+      initData,
+    } = encryptedEvent;
+    if (initData == null) {
+      const error = new Error("no init data found on media encrypted event.");
+      throw new EncryptedMediaError("INVALID_ENCRYPTED_EVENT", error, true);
+    }
+
+    const initDataBytes = new Uint8Array(initData);
+
+    if (handledInitData.has(initDataBytes, initDataType)) {
+      log.debug("init data already received. Skipping it.");
+      return Observable.empty(); // Already handled, quit
+    }
+    handledInitData.add(initDataBytes, initDataType);
+
+    // possible previous loaded session with the same initialization data
+    let previousLoadedSession : IMediaKeySession|MediaKeySession|null = null;
+    const { sessionsStore } = mediaKeysInfos;
+    const entry = sessionsStore.get(initDataBytes, initDataType);
+    if (entry != null) {
+      previousLoadedSession = entry.session;
+      if (isSessionUsable(previousLoadedSession)) {
+        log.debug("eme: reuse loaded session", previousLoadedSession.sessionId);
+        return Observable.of({
+          type: "loaded-open-session" as "loaded-open-session",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=212&issues=AWNFqgACXOBV_YrhBEpT&open=AWNFqgACXOBV_YrhBEpT)

[ISSUE_KEY]: AWNFqgACXOBV_YrhBEpT